### PR TITLE
Check if GC config exists before creating gc producer in replication

### PR DESCRIPTION
**Pull request by KillianG**
Issue: BB-569

### CODE REVIEW

File: extensions/replication/replicationStatusProcessor/ReplicationStatusProcessor.js [hunk 1]
```diff
@@ -370,8 +372,12 @@ class ReplicationStatusProcessor {
...
             },
             done => {
-                this._gcProducer = new GarbageCollectorProducer();
-                this._gcProducer.setupProducer(done);
+                if (this.enableGC) {
+                    this._gcProducer = new GarbageCollectorProducer();
+                    this._gcProducer.setupProducer(done);
+                } else {
+                    return done();
+                }
```
Comment (ghost) [inline extensions/replication/replicationStatusProcessor/ReplicationStatusProcessor.js]: For the linter

```suggestion
                if (this.enableGC) {
                    this._gcProducer = new GarbageCollectorProducer();
                    return this._gcProducer.setupProducer(done);
                }
                return done();
```

Comment (KillianG) [reply extensions/replication/replicationStatusProcessor/ReplicationStatusProcessor.js:380]: Linter doesnt seem unhappy, is it ?

File: extensions/replication/replicationStatusProcessor/ReplicationStatusProcessor.js [hunk 2]
```diff
@@ -204,14 +204,16 @@ class ReplicationStatusProcessor {
      *   in PEM format
      * @param {Object} mConfig - metrics config
      * @param {String} mConfig.topic - metrics config kafka topic
+     * @param {Boolean} enableGC - flag to enable garbage collection - Based on GC config
      */
     constructor(kafkaConfig, sourceConfig, repConfig,
-                internalHttpsConfig, mConfig) {
+                internalHttpsConfig, mConfig, enableGC) {
```
Comment (welansari) [inline extensions/replication/replicationStatusProcessor/ReplicationStatusProcessor.js]: Maybe we should just call it `gcConfig` ? `enableGC` is a bit missleading, it makes us think it's a boolean.

File: extensions/replication/replicationStatusProcessor/task.js [hunk 3]
```diff
@@ -13,11 +13,12 @@ const repConfig = config.extensions.replication;
 const sourceConfig = repConfig.source;
 const internalHttpsConfig = config.internalHttps;
 const mConfig = config.metrics;
+const gcEnabled = config.extensions.gc;
```
Comment (welansari) [inline extensions/replication/replicationStatusProcessor/task.js]: same here

File: extensions/replication/replicationStatusProcessor/ReplicationStatusProcessor.js [hunk 4]
```diff
@@ -370,8 +372,12 @@ class ReplicationStatusProcessor {
...
             },
             done => {
-                this._gcProducer = new GarbageCollectorProducer();
-                this._gcProducer.setupProducer(done);
+                if (this.gcConfig) {
+                    this._gcProducer = new GarbageCollectorProducer();
+                    return this._gcProducer.setupProducer(done);
+                } else {
+                    return done();
+                }
```
Comment (welansari) [inline extensions/replication/replicationStatusProcessor/ReplicationStatusProcessor.js]: Should add tests for this

File: tests/functional/replication/queueProcessor.js [hunk 5]
```diff
@@ -843,7 +843,7 @@ describe('queue processor functional tests with mocking', () => {
                           objectSizeMetrics: [100, 1000],
                         },
                         {},
-                        { topic: 'metrics-test-topic' });
+                        { topic: 'metrics-test-topic' }, { topic: 'backbeat-gc' });
```
Comment (ghost) [inline tests/functional/replication/queueProcessor.js]: Nit: consistency with other lines

```suggestion
                        { topic: 'metrics-test-topic' },
                        { topic: 'backbeat-gc' });
```

File: tests/unit/replication/ReplicationStatusProcessor.js [hunk 6]
```diff
@@ -12,7 +12,7 @@ function makeReplicationStatusProcessor(replayTopics) {
             transport: 'http',
         },
         { replayTopics, objectSizeMetrics: [100, 1000] },
-        {});
+        {}, false);
```
Comment (ghost) [inline tests/unit/replication/ReplicationStatusProcessor.js]:
```suggestion
        {},
        false);
```

File: extensions/replication/replicationStatusProcessor/ReplicationStatusProcessor.js [hunk 7]
```diff
@@ -370,8 +372,12 @@ class ReplicationStatusProcessor {
...
             },
```
Comment (welansari) [inline extensions/replication/replicationStatusProcessor/ReplicationStatusProcessor.js]:
```suggestion
                if (this.gcConfig) {
                    this._gcProducer = new GarbageCollectorProducer();
                    return this._gcProducer.setupProducer(done);
                }
                return done();
            },
```
nit

File: tests/functional/replication/queueProcessor.js [hunk 8]
```diff
@@ -1449,3 +1450,65 @@ describe('queue processor functional tests with mocking', () => {
...
+                  },
+                  groupId: 'backbeat-func-test-group-id',
+              },
+              monitorReplicationFailures: true,
+              objectSizeMetrics: [100, 1000],
+            },
+            {},
+            { topic: 'metrics-test-topic' },
+            { topic: 'backbeat-gc' });
+            rsp.start({ bootstrap: true }, done);
```
Comment (welansari) [inline tests/functional/replication/queueProcessor.js]: We pass `done` to `start()` here (same with the second test), not sure the next lines are executed ?
The assertion should be inside the `start()` callback.

File: extensions/replication/replicationStatusProcessor/ReplicationStatusProcessor.js [hunk 9]
```diff
@@ -458,8 +464,6 @@ class ReplicationStatusProcessor {
     processKafkaEntry(kafkaEntry, done) {
         const sourceEntry = QueueEntry.createFromKafkaEntry(kafkaEntry);
         if (sourceEntry.error) {
-            this.logger.error('error processing source entry',
-                { error: sourceEntry.error });
             return process.nextTick(() => done(errors.InternalError));
```
Comment (welansari) [inline extensions/replication/replicationStatusProcessor/ReplicationStatusProcessor.js]: why remove this ?

File: extensions/replication/replicationStatusProcessor/ReplicationStatusProcessor.js [hunk 10]
```diff
@@ -370,8 +372,12 @@ class ReplicationStatusProcessor {
...
             },
             done => {
-                this._gcProducer = new GarbageCollectorProducer();
-                this._gcProducer.setupProducer(done);
+                if (this.gcConfig) {
+                    this._gcProducer = new GarbageCollectorProducer();
+                    this._gcProducer.setupProducer(done);
+                } else {
+                    done();
+                }
```
Comment (welansari) [inline extensions/replication/replicationStatusProcessor/ReplicationStatusProcessor.js:380]: Looks like we never test this

**Review by welansari — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Pull request merged**